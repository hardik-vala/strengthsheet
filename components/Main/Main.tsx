import { StatusBar } from "expo-status-bar";
import { useState } from "react";
import { Alert, Button, Text, TextInput } from "react-native";
import { appendToGoogleSheet } from "../../services/sheetService";
import { SignOut } from "../SignOut/SignOut";

interface MainProps {
  onSignOut: () => void;
}

export function Main({ onSignOut }: MainProps) {
  const [userInput, setUserInput] = useState("");

  async function _appendToSheet() {
    const spreadsheetId = "1-wL-dRJYZkZ-uVpoBSuGeSFEzg_ZWVKFwLTv8RgbX7o";
    try {
      await appendToGoogleSheet(spreadsheetId, userInput);
    } catch (error) {
      Alert.alert(error);
    }
  }

  return (
    <>
      <Text>Enter a workout value:</Text>
      <StatusBar style="auto" />
      <TextInput
        style={{
          height: 40,
          width: "80%",
          borderColor: "gray",
          borderWidth: 1,
        }}
        defaultValue=""
        placeholder="Type here"
        value={userInput}
        onChangeText={(text) => setUserInput(text)}
      />
      <Button title="Submit" onPress={_appendToSheet} />
      <SignOut onSignOut={onSignOut} />
    </>
  );
}
