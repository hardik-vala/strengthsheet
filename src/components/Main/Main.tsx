import { format } from "date-fns";
import { StatusBar } from "expo-status-bar";
import { useState } from "react";
import { Alert, View } from "react-native";
import { Button, Divider } from "react-native-paper";
import { appendToGoogleSheet } from "../../services/sheetService";
import { DateTimePicker } from "../DateTimePicker/DateTimePicker";
import { FormInputField } from "../FormInputField/FormInputField";
import { SignOut } from "../SignOut/SignOut";

interface MainProps {
  onSignOut: () => void;
}

export function Main({ onSignOut }: MainProps) {
  const [dateInput, setDateInput] = useState(new Date());
  const [startTimeInput, setStartTimeInput] = useState(new Date());
  const [rowerWorkoutTimeInput, setRowerWorkoutTimeInput] = useState("");
  const [rowerWorkoutDistanceInput, setRowerWorkoutDistanceInput] =
    useState("");
  const [isAppendingToSheet, setIsAppendingToSheet] = useState(false);

  async function _appendToSheet() {
    const spreadsheetId = "1-wL-dRJYZkZ-uVpoBSuGeSFEzg_ZWVKFwLTv8RgbX7o";

    try {
      await appendToGoogleSheet(spreadsheetId, [
        format(dateInput, "MM/dd/yyyy"),
        format(startTimeInput, "HH:mm"),
        rowerWorkoutTimeInput,
        rowerWorkoutDistanceInput,
      ]);
    } catch (error) {
      Alert.alert("Error writing to Google Sheets");
      console.error(error);
      return;
    }
  }

  return (
    <View>
      <StatusBar style="auto" />
      <View
        style={{
          flexDirection: "row",
          justifyContent: "space-evenly",
        }}
      >
        <DateTimePicker
          label="Date"
          testID="datePicker"
          value={dateInput}
          isDate={true}
          onChange={(_, d) => setDateInput(d)}
        />
        <DateTimePicker
          label="Start time"
          testID="timePicker"
          value={startTimeInput}
          isDate={false}
          onChange={(_, t) => setStartTimeInput(t)}
        />
      </View>
      <Divider />
      <FormInputField
        label="Total Time"
        placeholder="30:00"
        value={rowerWorkoutTimeInput}
        onChangeText={(text) => setRowerWorkoutTimeInput(text)}
      />
      <Divider />
      <FormInputField
        label="Total Distance"
        placeholder="5000m"
        value={rowerWorkoutDistanceInput}
        onChangeText={(text) => setRowerWorkoutDistanceInput(text)}
      />
      <Divider />
      <View style={{ flexDirection: "row", justifyContent: "center" }}>
        <Button
          buttonColor="blue"
          compact={true}
          loading={isAppendingToSheet}
          mode="contained"
          onPress={async () => {
            setIsAppendingToSheet(true);
            await _appendToSheet();
            setIsAppendingToSheet(false);
          }}
          contentStyle={{
            flexDirection: "row",
          }}
          style={{
            marginTop: 25,
            height: 40,
            width: "25%",
            alignItems: "center",
            justifyContent: "center",
          }}
        >
          {isAppendingToSheet ? "Saving" : "Save"}
        </Button>
      </View>
      <SignOut onSignOut={onSignOut} />
    </View>
  );
}
