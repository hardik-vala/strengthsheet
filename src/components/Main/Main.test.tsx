import { act, fireEvent, render, screen } from "@testing-library/react-native";
import { format } from "date-fns";
import React from "react";

import { Main } from "./Main";

jest.useFakeTimers().setSystemTime(new Date("2023-08-27"));

jest.mock("../../services/sheetService", () => {
  return {
    appendToGoogleSheet: jest.fn(),
    readSheetValues: jest.fn(),
  };
});

import {
  appendToGoogleSheet,
  readSheetValues,
} from "../../services/sheetService";

const mockOnSignOut = jest.fn();

describe("Main", () => {
  // it("renders correctly", () => {
  //   const tree = render(<Main onSignOut={mockOnSignOut} />).toJSON();

  //   expect(tree).toMatchSnapshot();
  // });

  it("appends the text input to a spreadsheet upon submission", () => {
    render(<Main onSignOut={mockOnSignOut} />);
    
    act(() => {
      fireEvent.changeText(screen.getByPlaceholderText("30:00"), "foo");
      fireEvent.changeText(screen.getByPlaceholderText("5000m"), "bar");
      console.log(JSON.stringify(screen.toJSON()));
      // fireEvent.press(screen.getByText("Save"));
      // expect(appendToGoogleSheet).toHaveBeenCalledWith(
      //   "1-wL-dRJYZkZ-uVpoBSuGeSFEzg_ZWVKFwLTv8RgbX7o",
      //   [
      //     format(new Date(), "MM/dd/yyyy"),
      //     format(new Date(), "HH:mm"),
      //     "foo",
      //     "bar",
      //   ]
      // );
    });
  });
});
