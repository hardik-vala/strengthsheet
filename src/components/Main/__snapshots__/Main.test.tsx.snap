// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main renders correctly 1`] = `
<View>
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-evenly",
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "column",
          "padding": 10,
        }
      }
    >
      <Text
        style={
          [
            {
              "textAlign": "left",
            },
            {
              "color": "rgba(28, 27, 31, 1)",
              "writingDirection": "ltr",
            },
            [
              {
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": "400",
                "letterSpacing": 0.15,
                "lineHeight": 24,
              },
              undefined,
            ],
          ]
        }
      >
        Date
      </Text>
      <RNDateTimePicker
        date={1693094400000}
        displayIOS="default"
        enabled={true}
        is24Hour={true}
        mode="date"
        onChange={[Function]}
        onPickerDismiss={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "marginTop": 5,
          }
        }
        testID="datePicker"
      />
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "column",
          "padding": 10,
        }
      }
    >
      <Text
        style={
          [
            {
              "textAlign": "left",
            },
            {
              "color": "rgba(28, 27, 31, 1)",
              "writingDirection": "ltr",
            },
            [
              {
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": "400",
                "letterSpacing": 0.15,
                "lineHeight": 24,
              },
              undefined,
            ],
          ]
        }
      >
        Start time
      </Text>
      <RNDateTimePicker
        date={1693094400000}
        displayIOS="default"
        enabled={true}
        is24Hour={true}
        mode="time"
        onChange={[Function]}
        onPickerDismiss={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "marginTop": 5,
          }
        }
        testID="timePicker"
      />
    </View>
  </View>
  <View
    style={
      [
        {
          "backgroundColor": "rgba(202, 196, 208, 1)",
          "height": 0.5,
        },
        undefined,
        false,
        false,
        undefined,
      ]
    }
  />
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "padding": 10,
        "width": "100%",
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "column",
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          [
            {
              "textAlign": "left",
            },
            {
              "color": "rgba(28, 27, 31, 1)",
              "writingDirection": "ltr",
            },
            [
              {
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": "400",
                "letterSpacing": 0.15,
                "lineHeight": 24,
              },
              undefined,
            ],
          ]
        }
      >
        Total Time
      </Text>
    </View>
    <View
      style={
        [
          {
            "backgroundColor": "rgba(231, 224, 236, 1)",
            "borderTopLeftRadius": 4,
            "borderTopRightRadius": 4,
          },
          {
            "width": 100,
          },
        ]
      }
    >
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "rgba(28, 27, 31, 1)",
            "bottom": 0,
            "height": 1,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "transform": [
              {
                "scaleY": 0.5,
              },
            ],
            "zIndex": 1,
          }
        }
        testID="text-input-underline"
      />
      <View
        style={
          [
            {
              "paddingBottom": 0,
              "paddingTop": 0,
            },
            {
              "minHeight": 30,
            },
          ]
        }
      >
        <TextInput
          cursorColor="rgba(103, 80, 164, 1)"
          defaultValue=""
          editable={true}
          id="form-input-field"
          maxFontSizeMultiplier={1.5}
          multiline={false}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="30:00"
          placeholderTextColor="rgba(73, 69, 79, 1)"
          selectionColor="rgba(103, 80, 164, 1)"
          style={
            [
              {
                "margin": 0,
              },
              {
                "height": 30,
              },
              {
                "paddingBottom": 0,
                "paddingTop": 0,
              },
              {
                "color": "rgba(73, 69, 79, 1)",
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": undefined,
                "letterSpacing": 0.15,
                "lineHeight": undefined,
                "paddingLeft": 16,
                "paddingRight": 16,
                "textAlign": "right",
                "textAlignVertical": "center",
              },
              false,
              [
                {},
              ],
              undefined,
            ]
          }
          testID="text-input-flat"
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
    </View>
  </View>
  <View
    style={
      [
        {
          "backgroundColor": "rgba(202, 196, 208, 1)",
          "height": 0.5,
        },
        undefined,
        false,
        false,
        undefined,
      ]
    }
  />
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "padding": 10,
        "width": "100%",
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "column",
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          [
            {
              "textAlign": "left",
            },
            {
              "color": "rgba(28, 27, 31, 1)",
              "writingDirection": "ltr",
            },
            [
              {
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": "400",
                "letterSpacing": 0.15,
                "lineHeight": 24,
              },
              undefined,
            ],
          ]
        }
      >
        Total Distance
      </Text>
    </View>
    <View
      style={
        [
          {
            "backgroundColor": "rgba(231, 224, 236, 1)",
            "borderTopLeftRadius": 4,
            "borderTopRightRadius": 4,
          },
          {
            "width": 100,
          },
        ]
      }
    >
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "rgba(28, 27, 31, 1)",
            "bottom": 0,
            "height": 1,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "transform": [
              {
                "scaleY": 0.5,
              },
            ],
            "zIndex": 1,
          }
        }
        testID="text-input-underline"
      />
      <View
        style={
          [
            {
              "paddingBottom": 0,
              "paddingTop": 0,
            },
            {
              "minHeight": 30,
            },
          ]
        }
      >
        <TextInput
          cursorColor="rgba(103, 80, 164, 1)"
          defaultValue=""
          editable={true}
          id="form-input-field"
          maxFontSizeMultiplier={1.5}
          multiline={false}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="5000m"
          placeholderTextColor="rgba(73, 69, 79, 1)"
          selectionColor="rgba(103, 80, 164, 1)"
          style={
            [
              {
                "margin": 0,
              },
              {
                "height": 30,
              },
              {
                "paddingBottom": 0,
                "paddingTop": 0,
              },
              {
                "color": "rgba(73, 69, 79, 1)",
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": undefined,
                "letterSpacing": 0.15,
                "lineHeight": undefined,
                "paddingLeft": 16,
                "paddingRight": 16,
                "textAlign": "right",
                "textAlignVertical": "center",
              },
              false,
              [
                {},
              ],
              undefined,
            ]
          }
          testID="text-input-flat"
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
    </View>
  </View>
  <View
    style={
      [
        {
          "backgroundColor": "rgba(202, 196, 208, 1)",
          "height": 0.5,
        },
        undefined,
        false,
        false,
        undefined,
      ]
    }
  />
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "blue",
          "borderRadius": 20,
          "height": 40,
          "marginTop": 25,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0,
          "shadowRadius": 0,
          "width": "25%",
        }
      }
      testID="button-container-outer-layer"
    >
      <View
        collapsable={false}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "blue",
            "borderColor": "transparent",
            "borderRadius": 20,
            "borderStyle": "solid",
            "borderWidth": 0,
            "flex": 1,
            "justifyContent": "center",
            "minWidth": "auto",
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
            "shadowRadius": 0,
          }
        }
        testID="button-container"
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              {
                "overflow": "hidden",
              },
              {
                "borderRadius": 20,
              },
            ]
          }
          testID="button"
        >
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                [
                  {
                    "textAlign": "left",
                  },
                  {
                    "color": "rgba(28, 27, 31, 1)",
                    "writingDirection": "ltr",
                  },
                  [
                    {
                      "fontFamily": "System",
                      "fontSize": 14,
                      "fontWeight": "500",
                      "letterSpacing": 0.1,
                      "lineHeight": 20,
                    },
                    [
                      {
                        "marginHorizontal": 16,
                        "marginVertical": 9,
                        "textAlign": "center",
                      },
                      false,
                      {
                        "marginHorizontal": 24,
                        "marginVertical": 10,
                      },
                      {
                        "marginHorizontal": 8,
                      },
                      false,
                      {
                        "color": "rgba(255, 255, 255, 1)",
                        "fontFamily": "System",
                        "fontSize": 14,
                        "fontWeight": "500",
                        "letterSpacing": 0.1,
                        "lineHeight": 20,
                      },
                      undefined,
                    ],
                  ],
                ]
              }
              testID="button-text"
            >
              Save
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        [
          {},
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Log out
      </Text>
    </View>
  </View>
</View>
`;
